services/src/main/java/org/apache/unomi/services/impl/rules/RulesServiceImpl.java

```diff
@@ -393,6 +408,11 @@ public void setRule(Rule rule) {
                 definitionsService.extractConditionBySystemTag(condition, "eventCondition");
             }
         }
+        List<Action> actions = rule.getActions();
+        if (actions == null || actions.isEmpty()) {
+            logger.warn("rule {} won't be saved as it contains no action", rule.getItemId());
+            return;
+        }
```

Comment:
We need to check if the rule is disabled here, in case it is disabled we could authorize the persitence even if there is no action.
It's like a draft rule.

Also it could be interesting to force the disable but save anyway in case of no action to avoid user loosing there work in case they are using some sort of UI to build the rule, they may not be aware that the rule is not saved at the end.

Also in case we still want to not save the rule, then we should crash an Exception or a validation error message. Because the setRule didn't do the expected behavior.
I think force disabling the rule and save anyway would be the more easy to do. and more logic considering the rest of the fix

Reply:
I think I would be more in favor of raising an exception in this case.